app/src/main/java/mod/elfilibustero/sketch/lib/valid/VariableTypeValidator.java

```diff
@@ -0,0 +1,40 @@
+package mod.elfilibustero.sketch.lib.valid;
+
+import a.a.a.MB;
+import android.content.Context;
+import android.text.TextUtils;
+import java.util.regex.Pattern;
+import com.google.android.material.textfield.TextInputLayout;
+
+public class VariableTypeValidator extends MB {
+    public static final Pattern PATTERN_TYPE = Pattern.compile("^([a-zA-Z0-9]+)(<([a-zA-Z<>]+(?:\\s*,(?!.*?,)\\s*[a-zA-Z<>]+)*)>)?(\\[\\])?$");
```

Comment:
generics still don't allow numbers, and `$` isn't allowed at all

Reply:
also e.g. `List<?>`, `List<? extends Number>`, and `List<? super Number>` don't work

Reply:
that's the reason I suggested raw input only, relying on the user to know what they're doing